torch/csrc/distributed/c10d/nvshmem_extension.cu

```diff
@@ -344,20 +350,82 @@ __global__ void exchangeSplitAndOffset_2d(int64_t* in_out_splits, int mype, int
+  __shared__ typename WarpScan::TempStorage temp_storage[NUM_WARPS];
+
+  // Obtain input item for each thread
+  int tid = threadIdx.x % WARP_SIZE;
+  int64_t thread_data = (tid < n) ? idata[tid] : 0;
+
+  // Compute the warp-wide exclusive prefix sum
+  int warp_id = threadIdx.x / WARP_SIZE;
+  WarpScan(temp_storage[warp_id]).ExclusiveSum(thread_data, thread_data);
+
+  // Store the result
+  odata[tid] = thread_data;
+}
+
+// This is for abstracting a thread-group-scope, exclusive prefix sum.
+// Since we use warp-scope prefix sum, the thread group size is limited to warp size.
+#define A2AV_TILE_SIZE WARP_SIZE
+
 // This kernel is used to do the actual data exchange.
 // `in_out_splits` has the same definition as in `exchangeSplitAndOffset`.
 // `stride` is the stride at dim 0, unit in byte.
 // For meaning of `mype` and `npes`, see the docstring of `nvshmem_all_to_all_vdev_2d`.
-__global__ void allToAllV_2d(void *send_data, void *recv_data, int64_t* in_out_splits, size_t stride, int mype, int npes, int ne) {
+__global__ void allToAllV_2d(void *send_data, void *recv_data, int64_t* in_out_splits, size_t stride, int mype, int npes, int ne, int64_t major_align) {
   int nsplits = npes * ne;
   auto output_splits = in_out_splits + nsplits;
   auto source_offsets = in_out_splits + nsplits * 2;
   int bid = blockIdx.x;
   int tid = threadIdx.x;
 
-  // Calculate the output offsets
-  __shared__ int64_t e_offsets[THREADS_PER_BLOCK];
-  prefixSum(e_offsets, output_splits, nsplits);
+  // Split the thread block into tiles
+  constexpr int NUM_TILES = THREADS_PER_BLOCK / A2AV_TILE_SIZE;
+  thread_group tile = cg::tiled_partition(this_thread_block(), A2AV_TILE_SIZE);
+  int tileId = tid / A2AV_TILE_SIZE;
+  // Each tile calculates its own prefix sum
+  __shared__ int64_t tile_prefix_sums[NUM_TILES][A2AV_TILE_SIZE];
+  // A tile takes care of npes worth of splits
+  int nsplits_per_tile = min(npes, nsplits - tileId * npes);
+  // TODO: currently it is assumed that the number of PE's is smaller than `A2AV_TILE_SIZE`
+  CUDA_KERNEL_ASSERT(nsplits_per_tile <= A2AV_TILE_SIZE);
+
+  // Total length of each tile
+  __shared__ int64_t len_per_tile[NUM_TILES];
+  // Starting offset of each tile
+  __shared__ int64_t start_offset_per_tile[NUM_TILES];
+  // This tile does not need to do tile-wise prefix sum
+  if (nsplits_per_tile < 0) goto end_of_tile_prefix_sum;
+
+  // Each tile calculates its own prefix sum
+  prefixSum_warp(tile_prefix_sums[tileId], output_splits + tileId * npes, nsplits_per_tile);
+
+  // Last thread in each tile does the up aligning.
+  // Note: using the last thread to read the last sum from `tile_prefix_sums` so
+  // that we can save a __syncthreads(). This is safe because the last thread is
+  // the one that writes the last sum in the prefixSum function.
+  if (tile.thread_rank() == A2AV_TILE_SIZE - 1) {
+    auto my_tile_len = tile_prefix_sums[tileId][A2AV_TILE_SIZE - 1] + output_splits[tileId * npes + nsplits_per_tile - 1];
```

Comment:
you should have computed it in prefixSum_warp using an overload with warp_aggregate https://nvidia.github.io/cccl/cub/api/classcub_1_1WarpScan.html#_CPPv4N3cub8WarpScan12ExclusiveSumE1TR1TR1T

Reply:
Nice!